x509/cert.go

```diff
@@ -30,3 +31,35 @@ func ParseCertificatePEM(data []byte) ([]*x509.Certificate, error) {
 	}
 	return certs, nil
 }
+
+// ValidateCertChain takes an ordered certificate chain and validates issuance from leaf to root
+func ValidateCertChain(certChain []*x509.Certificate) error {
+	if len(certChain) < 2 {
+		return errors.New("certificate chain must contain at least two certificates")
+	}
+
+	for i, cert := range certChain {
+		if i == len(certChain)-1 {
+			if !isSelfSigned(cert) {
+				return errors.New("certificate chain must end with a root certificate (root certificates are self-signed)")
+			}
+		} else {
+			if isSelfSigned(cert) {
+				return errors.New("certificate chain must not contain self-signed intermediates")
+			} else if nextCert := certChain[i+1]; !isIssuedBy(cert, nextCert) {
+				return errors.New("signature on certificate '" + nextCert.Subject.String() + "' is not issued by '" + cert.Issuer.String() + "'")
+			}
+		}
+	}
+
+	return nil
+}
+
+func isSelfSigned(cert *x509.Certificate) bool {
+	return isIssuedBy(cert, cert)
+}
+
+func isIssuedBy(subject *x509.Certificate, issuer *x509.Certificate) bool {
+	err := subject.CheckSignatureFrom(issuer)
+	return err == nil && issuer.Subject.String() == subject.Issuer.String()
```

Comment:
DNs are structures and shouldn't be treated as strings. See https://frasertweedale.github.io/blog-redhat/posts/2019-05-28-a-dn-is-not-a-string.html

Can we use Golang's in-built [chain validation](https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L749)?

Reply:
I actually originally wrote chain validation using Go's Verify method, but it's not ideal because it's intended for TLS certs. The code ends up being [2-3 times longer](https://paste.amazon.com/show/jddonas/1653497401) than what I have here, you have to specify a "current time" for the certs even if you want expired certs to still validate, and you have to know ahead of time the exact roots you want to verify against (you can't just validate up to "any" root)

In terms of the DN, I'll look into a better way to check equality. My original reasoning is that checking string equality should at least be consistent because I'm not sure how consistent struct equality is in Go. Will get back about this

Reply:
Fixed this by copying how the Verify method checks this - comparing raw Subject bytes to Issues bytes: [source](https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L565)